File: EvolutionaryGenerator_v2.1_Python/hermes/problems/kp/generator/kp_individual.py
"""Translation of ``mx.tec.hermes.problems.kp.generator.KPIndividual``."""

import math

from hermes.problems.kp.item import Item
from hermes.problems.kp.kp import KP
from javacompat import BitSet
from metah.individual import Individual


class KPIndividual(Individual):
    """Provides the methods to create and use individuals that encode knapsack problems
    that can be evolved by a genetic algorithm.

    :author: Paola Azeneth Castillo Gutiérrez (Python translation)
    :original author: José Carlos Ortiz Bayliss (jcobayliss@tec.mx)
    :original version: 2.0
    """

    # These four values are static in the Java version: every individual created
    # within the same run shares the same encoding.
    max_weight_per_item = 20
    max_profit_per_item = 50
    nb_items = 10
    capacity = 20
    nb_bits = 10 * int(math.ceil(math.log10(20) / math.log10(2)) + math.ceil(math.log10(50) / math.log10(2)))

    @classmethod
    def set_capacity(cls, capacity: int) -> None:
        """Sets the capacity of the knapsack in the resulting knapsack problem.

        :param capacity: The capacity of the knapsack in the resulting knapsack problem.
        """
        KPIndividual.capacity = capacity

    @classmethod
    def set_max_weight_per_item(cls, max_weight_per_item: int) -> None:
        """Sets the maximum weight per item in the resulting knapsack problem.

        :param max_weight_per_item: The maximum weight per item in the resulting knapsack problem.
        """
        KPIndividual.max_weight_per_item = max_weight_per_item
        KPIndividual.nb_bits = KPIndividual.nb_items * int(
            math.ceil(math.log10(KPIndividual.max_weight_per_item) / math.log10(2))
            + math.ceil(math.log10(KPIndividual.max_profit_per_item) / math.log10(2)))

    @classmethod
    def set_max_profit_per_item(cls, max_profit_per_item: int) -> None:
        """Sets the maximum profit per item in the resulting knapsack problem.

        :param max_profit_per_item: The maximum profit per item in the resulting knapsack problem.
        """
        KPIndividual.max_profit_per_item = max_profit_per_item
        KPIndividual.nb_bits = KPIndividual.nb_items * int(
            math.ceil(math.log10(KPIndividual.max_weight_per_item) / math.log10(2))
            + math.ceil(math.log10(KPIndividual.max_profit_per_item) / math.log10(2)))

    @classmethod
    def set_nb_items(cls, nb_items: int) -> None:
        """Sets the number of items in the resulting knapsack problem.

        :param nb_items: The number of items in the resulting knapsack problem.
        """
        KPIndividual.nb_items = nb_items
        KPIndividual.nb_bits = KPIndividual.nb_items * int(
            math.ceil(math.log10(KPIndividual.max_weight_per_item) / math.log10(2))
            + math.ceil(math.log10(KPIndividual.max_profit_per_item) / math.log10(2)))

    def __init__(self, seed: int = None, individual: "KPIndividual" = None):
        """Mirrors the two constructors available in the Java version.

        ``KPIndividual(seed)`` creates a new random individual while
        ``KPIndividual(individual=...)`` is the (private) copy constructor.

        :param seed: The seed to initialize the random number generator.
        :param individual: The instance of ``KPIndividual`` to copy.
        """
        if individual is not None:
            super().__init__(individual.get_evaluation(), individual.random.next_long())
            self.chromosome = individual.chromosome.clone()
        else:
            super().__init__(0, seed)
            self.chromosome = BitSet(KPIndividual.nb_bits)
            for i in range(KPIndividual.nb_bits):
                self.chromosome.set(i, self.random.next_boolean())

    def combine(self, individual: Individual, crossover_rate: float) -> list:
        offspring = [self, individual]
        crossover_point = self.random.next_int(KPIndividual.nb_bits)
        tmp = offspring[1].chromosome.clone()
        for i in range(crossover_point):
            offspring[1].chromosome.set(i, offspring[0].chromosome.get(i))
        for i in range(crossover_point):
            offspring[0].chromosome.set(i, tmp.get(i))
        return offspring

    def mutate(self, mutation_rate: float) -> None:
        for i in range(KPIndividual.nb_bits):
            if self.random.next_double() < mutation_rate:
                self.chromosome.flip(i)

    def copy(self) -> Individual:
        return KPIndividual(individual=self)

    def to_kp(self) -> KP:
        """Returns a new instance of ``KP`` based on the information contained in this individual.

        :return: A new instance of ``KP``.
        """
        items = []
        bits_block = int(math.ceil(math.log10(KPIndividual.max_weight_per_item) / math.log10(2))
                         + math.ceil(math.log10(KPIndividual.max_profit_per_item) / math.log10(2)))
        bits_weight = int(math.ceil(math.log10(KPIndividual.max_weight_per_item) / math.log10(2)))
        bits_profit = int(math.ceil(math.log10(KPIndividual.max_profit_per_item) / math.log10(2)))
        for i in range(KPIndividual.nb_items):
            from_index = i * bits_block
            bits = self.chromosome.get(from_index, from_index + bits_weight)
            weight = KPIndividual.to_integer(bits)
            from_index = from_index + bits_weight
            bits = self.chromosome.get(from_index, from_index + bits_profit)
            profit = KPIndividual.to_integer(bits)
            # hack, borrar!  ["hack, delete!" -- commented out block left by the author
            # in the Java original, kept verbatim.]
            #
            # profit = int((profit + 1) / 128.0 * 100)
            # if profit == 0 or profit > 100:
            #     print("Out of range generation.")
            #     print(profit)
            #     sys.exit(1)
            #
            items.append(Item(i, profit, weight))
        return KP(items, KPIndividual.capacity)

    def __str__(self) -> str:
        string = []
        bits_block = int(math.ceil(math.log10(KPIndividual.max_weight_per_item) / math.log10(2))
                         + math.ceil(math.log10(KPIndividual.max_profit_per_item) / math.log10(2)))
        bits_weight = int(math.ceil(math.log10(KPIndividual.max_weight_per_item) / math.log10(2)))
        bits_profit = int(math.ceil(math.log10(KPIndividual.max_profit_per_item) / math.log10(2)))
        for i in range(KPIndividual.nb_items):
            from_index = i * bits_block
            bits = self.chromosome.get(from_index, from_index + bits_weight)
            tmp_string = []
            for j in range(bits_weight):
                tmp_string.append("1" if bits.get(j) else "0")
            string.append("".join(reversed(tmp_string)) + " (" + str(KPIndividual.to_integer(bits)) + ") ")
            from_index = from_index + bits_weight
            bits = self.chromosome.get(from_index, from_index + bits_profit)
            tmp_string = []
            for j in range(bits_profit):
                # The Java version reads bits.get(j + from) here, which is out of the
                # range of the extracted bits.  It is kept as it is on purpose.
                tmp_string.append("1" if bits.get(j + from_index) else "0")
            string.append("".join(reversed(tmp_string)) + " (" + str(KPIndividual.to_integer(bits)) + ") ")
        return "".join(string)

    @staticmethod
    def to_integer(bits: BitSet) -> int:
        """Returns the value as integer of the bits provided as argument.

        :param bits: The bits whose integer value is required.
        :return: The value as integer of the bits provided as argument.
        """
        value = 0
        i = bits.next_set_bit(0)
        while i >= 0:
            value += int(math.pow(2, i))
            i = bits.next_set_bit(i + 1)
        return value + 1

File: EvolutionaryGenerator_v2.1_Java/Generator/src/mx/tec/hermes/problems/kp/generator/KPIndividual.java
package mx.tec.hermes.problems.kp.generator;

import mx.tec.hermes.problems.kp.Item;
import java.util.ArrayList;
import java.util.BitSet;
import java.util.List;
import mx.tec.hermes.problems.kp.KP;
import mx.tec.meta.Individual;

/**
 * Provides the methods to create and use individuals that encode knapsack problems that can be evolved by a genetic
 * algorithm.
 *
 * @author José Carlos Ortiz Bayliss (jcobayliss@tec.mx)
 * @version 2.0
 */
public class KPIndividual extends Individual {

    protected static int maxWeightPerItem = 20, maxProfitPerItem = 50, nbItems = 10, capacity = 20;
    protected static int nbBits = nbItems * (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)) + Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
    protected final BitSet chromosome;

    /**
     * Sets the capacity of the knapsack in the resulting knapsack problem.
     *
     * @param capacity The capacity of the knapsack in the resulting knapsack problem.
     */
    public static void setCapacity(int capacity) {
        KPIndividual.capacity = capacity;
    }

    /**
     * Sets the maximum weight per item in the resulting knapsack problem.
     *
     * @param maxWeightPerItem The maximum weight per item in the resulting knapsack problem.
     */
    public static void setMaxWeightPerItem(int maxWeightPerItem) {
        KPIndividual.maxWeightPerItem = maxWeightPerItem;
        nbBits = nbItems * (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)) + Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
    }

    /**
     * Sets the maximum profit per item in the resulting knapsack problem.
     *
     * @param maxProfitPerItem The maximum profit per item in the resulting knapsack problem.
     */
    public static void setMaxProfitPerItem(int maxProfitPerItem) {
        KPIndividual.maxProfitPerItem = maxProfitPerItem;
        nbBits = nbItems * (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)) + Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
    }

    /**
     * Sets the number of items in the resulting knapsack problem.
     *
     * @param nbItems The number of items in the resulting knapsack problem.
     */
    public static void setNbItems(int nbItems) {
        KPIndividual.nbItems = nbItems;
        nbBits = nbItems * (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)) + Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
    }

    /**
     * Creates a new instance of <code>KPIndividual</code>.
     * 
     * @param seed The seed to initialize the random number generator.
     */
    public KPIndividual(long seed) {
        super(0, seed);
        chromosome = new BitSet(nbBits);
        for (int i = 0; i < nbBits; i++) {
            chromosome.set(i, random.nextBoolean());
        }
    }

    @Override
    public Individual[] combine(Individual individual, double crossoverRate) {
        int crossoverPoint;
        BitSet tmp;
        KPIndividual[] offspring;
        offspring = new KPIndividual[]{(KPIndividual) this, (KPIndividual) individual};
        crossoverPoint = random.nextInt(nbBits);
        tmp = (BitSet) offspring[1].chromosome.clone();
        for (int i = 0; i < crossoverPoint; i++) {
            offspring[1].chromosome.set(i, offspring[0].chromosome.get(i));
        }
        for (int i = 0; i < crossoverPoint; i++) {
            offspring[0].chromosome.set(i, tmp.get(i));
        }
        return offspring;
    }

    @Override
    public void mutate(double mutationRate) {
        for (int i = 0; i < nbBits; i++) {
            if (random.nextDouble() < mutationRate) {
                chromosome.flip(i);
            }
        }
    }

    @Override
    public Individual copy() {
        return new KPIndividual(this);
    }

    /**
     * Returns a new instance of <code>KP</code> based on the information contained in this individual.
     *
     * @return A new instance of <code>KP</code>.
     */
    public KP toKP() {
        int profit, weight, bitsBlock, bitsWeight, bitsProfit, from;
        BitSet bits;
        List<Item> items;
        items = new ArrayList(nbItems);
        bitsBlock = (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)) + Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
        bitsWeight = (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)));
        bitsProfit = (int) (Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
        for (int i = 0; i < nbItems; i++) {
            from = i * bitsBlock;
            bits = chromosome.get(from, from + bitsWeight);
            weight = toInteger(bits);
            from = from + bitsWeight;
            bits = chromosome.get(from, from + bitsProfit);
            profit = toInteger(bits);
            // hack, borrar!
            /*
            profit = (int) ((profit + 1) / 128.0 * 100);
            if (profit == 0 || profit > 100) {
                System.out.println("Out of range generation.");
                System.out.println(profit);
                System.exit(1);
            }*/
            items.add(new Item(i, profit, weight));
        }
        return new KP(items, capacity);
    }
    
    @Override
    public String toString() {
        int bitsBlock, bitsWeight, bitsProfit, from;
        BitSet bits;
        StringBuilder string, tmpString;
        string = new StringBuilder();
        bitsBlock = (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)) + Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
        bitsWeight = (int) (Math.ceil(Math.log10(maxWeightPerItem) / Math.log10(2)));
        bitsProfit = (int) (Math.ceil(Math.log10(maxProfitPerItem) / Math.log10(2)));
        for (int i = 0; i < nbItems; i++) {
            from = i * bitsBlock;
            bits = chromosome.get(from, from + bitsWeight);
            tmpString = new StringBuilder();
            for (int j = 0; j < bitsWeight; j++) {
                tmpString.append(bits.get(j) ? "1" : "0");
            }
            string.append(tmpString.reverse().toString()).append(" (").append(toInteger(bits)).append(") ");
            from = from + bitsWeight;
            bits = chromosome.get(from, from + bitsProfit);
            tmpString = new StringBuilder();
            for (int j = 0; j < bitsProfit; j++) {
                tmpString.append(bits.get(j + from) ? "1" : "0");
            }
            string.append(tmpString.reverse().toString()).append(" (").append(toInteger(bits)).append(") ");

        }
        return string.toString();
    }

    /**
     * Creates a new instance of <code>KPIndividual</code>. This is a copy constructor.
     *
     * @param individual The instance of <code>KPIndividual</code> to copy.
     */
    private KPIndividual(KPIndividual individual) {
        super(individual.getEvaluation(), individual.random.nextLong());
        this.chromosome = (BitSet) individual.chromosome.clone();
    }

    /**
     * Returns the value as integer of the bits provided as argument.
     *
     * @param bits The bits whose integer value is required.
     * @return The value as integer of the bits provided as argument.
     */
    protected static int toInteger(BitSet bits) {
        int i, value;
        value = 0;
        i = bits.nextSetBit(0);
        while (i >= 0) {
            value += Math.pow(2, i);
            i = bits.nextSetBit(i + 1);
        }
        return value + 1;
    }

}
